[Feat] #9 - ProductDetailCompositionLayout, ProductRelatedCollectionViewCell 구현

choiyoubin opened the pull request:
## 🍽️ 작업 내용
<!-- 작업한 내용을 간단하게 적어주세요! -->
- Compositional Layout 구현
- ClassName 프로토콜 구현
- ProductRelatedCollectionViewCell 구현


## 🚗 PR Point
<!-- 주의할 사항이나 같이 고민해볼 부분, 강조하고 싶은 내용 등을 적어주세요! -->
- ClassName 프로토콜 사용법
```
extension ProductRelatedCollectionViewCell: ClassNameProtocol { }
extension으로 사용 가능
```

- Compositional Layout Switch 문으로 구현, Layout은 extension에 함수로 구현 

```
switch section {
   case .productImage:
          return self.productImageSectionLayout()
```
 
```
static private func productImageSectionLayout() -> NSCollectionLayoutSection {
        
        let item = NSCollectionLayoutItem(
            layoutSize: .init(
                widthDimension: .fractionalWidth(1),
                heightDimension: .fractionalHeight(1)
            )
        )
        
        let group = NSCollectionLayoutGroup.horizontal(
            layoutSize: .init(
                widthDimension: .fractionalWidth(1),
                heightDimension: .estimated(375)
            ),
            subitems: [item]
        )
        
        let section = NSCollectionLayoutSection(group: group)
        section.orthogonalScrollingBehavior = .paging
        
        return section
    } 
```


## 📱 스크린샷

![Simulator Screen Recording - iPhone 13 mini - 2024-11-21 at 03 41 46](https://github.com/user-attachments/assets/ff1ce2f8-37a8-4f45-afce-b3187b94f268)



## ✅ CheckList
- [x] 오류 없이 빌드되는지 확인
- [x] 로그용 print문 제거
- [x] 불필요한 주석 제거
- 코드 컨벤션 확인!

### 🔗 Issue 
<!-- 생성한 관련 이슈가 있다면 Resolved #이슈번호로 닫아주세요! -->
Resolved #9

choiyoubin committed:
[Feat/#9] ProductRelatedCollectionViewCell 구현

choiyoubin committed:
Merge remote-tracking branch 'origin/main' into Feat/#9-ProductRelatedCell
[Feat/#9] main 브랜치 pull

choiyoubin committed:
[Feat/#9] ProductDetailCompositionalLayout 구현 진행 중

ProductDetailViewController, ProductDetailView, ProductDetailSection 생성, ProductRelatedModel 내 샘플 데이터 생성

choiyoubin committed:
[Feat/#9] ProductDetailSection 수정 및 CompositionalLayout 수정 중

2x2 Grid 구현 위해 item 개수 4로 수정

choiyoubin committed:
[Feat/#9] ProductRelatedModel 목업데이터 수정 및 CompositionalLayout 수정

Compositional Layout 섹션별 설정

choiyoubin committed:
[Feat/#9] CompositionLayout 구현 및 ProductDetailViewController 구현

ProductRelatedCell을 제외한 나머지 Cell들은 샘플 cell

Nya128 commented on Daangn_iOS/Daangn_iOS/Global/Extensions/UICollectionViewCell+.swift:
```diff
@@ -0,0 +1,15 @@
+
+import UIKit
+
+extension UICollectionViewCell {
+    
+    static var className: String {
+        return String(describing: self)
+    }
+}
```
최고. 테이블뷰셀도 만들어주세요 ~ >__<

choiyoubin replied:
UITableViewCell+ 도 만들어 놓을게요!!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/KeywordNotifyCollectionViewCell.swift:
```diff
@@ -0,0 +1,52 @@
+    
+    private func setLayout() {
+        sampleView.snp.makeConstraints {
+            $0.size.equalToSuperview()
```
이럴땐 size보다는 edges가 더 적합해보입니다!!

choiyoubin replied:
헉 그렇네요.. 다음부턴 주의해서 사용할게요 ㅎㅎ sample로 만들어놨다고 생각없이 사용한 것 같습니다..ㅎㅎ

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/ProductRelatedCollectionViewCell.swift:
```diff
@@ -0,0 +1,93 @@
+    // MARK: - UI Components
+    
+    private let relatedSalesImageView = UIImageView()
+    private let relatedSalesLabelStackView = UIStackView()
```
스택뷰를 사용하신 이유가 있으실까요?! 
간단한 뷰라 스택뷰 없어도 충분히 구현 가능할 것 같아서(성능적으로도 더 나을듯)
스택뷰 빼고 이미지뷰, 라벨 2개로 레이아웃 잡는 게 좋을 거 같아요!

choiyoubin replied:
넵 레이아웃 확인하다가 변경하지 못하였네요!

choiyoubin replied:
변경해서 push하겠습니다~!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/ProductRelatedCollectionViewCell.swift:
```diff
@@ -0,0 +1,93 @@
+    }
+    
+    private func setHierarchy() {
+        relatedSalesLabelStackView.addArrangeSubViews(
+            relatedSalesTitleLabel, relatedSalesPriceLabel
+        )
+        
+        contentView.addSubviews(
+            relatedSalesImageView, relatedSalesLabelStackView
+        )
```
파라미터를 따로 쓰는 부분이 없어서 줄바꿈 안해도 괜찮을듯ㅎㅎ

choiyoubin replied:
오 다음부터 컨벤션 인지하고 사용하겠습니다~ 수정해서 반영할게요!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/ProductRelatedCollectionViewCell.swift:
```diff
@@ -0,0 +1,93 @@
+    
+    private func setLayout() {
+        relatedSalesImageView.snp.makeConstraints {
+            $0.horizontalEdges.equalToSuperview()
```
제약조건을 잡을 때는 반드시 x축, y축 하나씩은 지정해줘야 정확하게 나옵니다!
이미지뷰의 top도 equalToSuperview()로 잡아주세용

choiyoubin replied:
넵 수정하겠습니다!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/ProductRelatedCollectionViewCell.swift:
```diff
@@ -0,0 +1,93 @@
+        relatedSalesLabelStackView.snp.makeConstraints {
+            $0.horizontalEdges.equalToSuperview()
+            $0.top.equalTo(relatedSalesImageView.snp.bottom).offset(8)
+        }
```
Cell이 내부 컨텐츠대로 높이를 계산하려면 bottom 제약조건도 필요하니 꼭 넣어주세요~!~!

choiyoubin replied:
$0.bottom.equalToSuperView()를 relatedSalesPriceLabel에 추가하여 Cell 이 내부 컨텐츠대로 높이를 계산할 수 있도록 수정하였습니다

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Cells/ProductRelatedHeaderCell.swift:
```diff
@@ -0,0 +1,62 @@
+import SnapKit
+import Then
+
+class ProductRelatedHeaderCell: UICollectionViewCell {
```
헤더는 UICollectionReusableView로 만들어주시면 됩니다!

choiyoubin replied:
변경 하였습니다!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/ProductDetailCompositionalLayout.swift:
```diff
@@ -0,0 +1,172 @@
+            
+            let section: NSCollectionLayoutSection
+            
+            if sectionNumber == 0 {
+                
+                let item = NSCollectionLayoutItem(
+                    layoutSize: .init(
+                        widthDimension: .fractionalWidth(1),
+                        heightDimension: .fractionalHeight(1)
+                    )
+                )
+                
+                let group = NSCollectionLayoutGroup.horizontal(
+                    layoutSize: .init(
+                        widthDimension: .fractionalWidth(1),
+                        heightDimension: .estimated(375)
+                    ),
+                    subitems: [item]
+                )
+                
+                section = NSCollectionLayoutSection(group: group)
+                section.orthogonalScrollingBehavior = .paging
+                
+            } else if sectionNumber == 1 {
```
if문보다는 switch문 쓰는 게 더 깔끔할 듯합니다.
그리고 각각의 섹션별로 레이아웃 설정 함수를 만들면 코드 읽기 편할 거 같아요~(아래 코드 참고)
추가로 0, 1, 2 Int값으로 설정해도 좋지만 enum case 만들어둔대로 .productImage 이런식으로 스위치문 써서 구분하는 것도 좋아보입니다 ㅎㅎ
```
switch sectionNumber {
case .productImage:
    return self.productImageSectionLayout()
case .sellerInfo:
    return self.sellerInfoSectionLayout()
case .productDetailInfo:
    return self.productDetailInfoSectionLayout()
...
}
```
```
private func productImageSectionLayout() {
    let item = NSCollectionLayoutItem(
          layoutSize: .init(
              widthDimension: .fractionalWidth(1),
              heightDimension: .fractionalHeight(1)
          )
      )

      let group = NSCollectionLayoutGroup.horizontal(
          layoutSize: .init(
              widthDimension: .fractionalWidth(1),
              heightDimension: .estimated(375)
          ),
          subitems: [item]
      )

      section = NSCollectionLayoutSection(group: group)
      section.orthogonalScrollingBehavior = .paging
}
```

choiyoubin replied:
수정했는데 확인 한번 부탁드려용~

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/ViewControllers/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,191 @@
+    
+    // MARK: - View Life Cycle
+    
+    override func loadView() {
```
super.loadView()도 붙여주세용 ㅎㅎ

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/ViewControllers/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,191 @@
+        _ collectionView: UICollectionView,
+        numberOfItemsInSection section: Int
+    ) -> Int {
+        guard let section = ProductDetailSection(rawValue: section)
+        else {
+            return 0
+        }
```
guard let은 줄바꿈 굳이 안 해도 괜찮다고 생각합니다 😆

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/ViewControllers/ProductDetailViewController.swift:
```diff
@@ -0,0 +1,191 @@
+            let cell = collectionView.dequeueReusableCell(
+                withReuseIdentifier: ProductImageCollectionViewCell.className,
+                for: indexPath
+            ) as! ProductImageCollectionViewCell
```
guard let을 써서 as?로 타입 캐스팅하면 더 안전성이 높아지지 않을까 싶습니당

choiyoubin replied:
guard let 으로 캐스팅에 실패했을때 기본 UICollectionViewCell()을 반환하도록 수정하였습니다!

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Views/ProductDetailView.swift:
```diff
@@ -0,0 +1,52 @@
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        backgroundColor = .white
```
이것도 setStyle()에 넣어주시면 됩니당 ㅎ!!

choiyoubin replied:
어라 이거 왜 여기다 뺴먹었지..

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/Views/ProductDetailView.swift:
```diff
@@ -0,0 +1,52 @@
+        addSubview(collectionView)
+    }
+    
+    private func setLayout() {
+        collectionView.snp.makeConstraints {
+            $0.edges.equalToSuperview()
+        }
+    }
```
goooood 👍

Nya128 reviewed (approved):
전체적으로 깔끔하고 좋아요~!~! 고생하셨습니다. 👍☺️
코멘트 남긴거 다 반영하면 제가 확인한 후 머지하는 걸로!!

choiyoubin committed:
[Feat/#9] UITableView+ Extension 구현

className 접근하는 extension

choiyoubin committed:
[Chore/#9] Cells Layout 수정

코드리뷰 반영

choiyoubin committed:
[Chore/#9] StackView 삭제 후 ImageView, Label 간 레이아웃 작성

choiyoubin committed:
[Chore/#9] guard let 사용하여 타입 캐스팅

+ ProductDetailView setStyle()에 background 색상 추가 / ViewController super.loadView() 추가

choiyoubin committed:
[Feat/#9] ClassNameProtocol 구현 및 사용 , Header 영역 ReusableView로 변경

+ ProductRelatedCollectionViewCell bottom layout 추가

choiyoubin committed:
[Refactor/#9] Compositional Layout Switch문 변경, 섹션별 레이아웃 함수 구현

Nya128 commented on Daangn_iOS/Daangn_iOS/Presentation/Product/ProductDetailCompositionalLayout.swift:
```diff
@@ -0,0 +1,212 @@
+            guard let section = ProductDetailSection(rawValue: sectionNumber)
+            else { fatalError() }
+            
+            switch section {
+            case .productImage:
+                return self.productImageSectionLayout()
+            case .sellerInfo:
+                return self.sellerInfoSectionLayout()
+            case .productDetailInfo:
+                return self.productDetailInfoSectionLayout()
+            case .sellerProduct:
+                return self.sellerProductSectionLayout()
+            case .keywordNotify:
+                return self.keywordNotifySectionLayout()
+            case .relatedArticle:
+                return self.relatedArticleSectionLayout()
+            }
```
so 깔끔. 👍

Nya128 reviewed (commented):
확인 완료. 머지하셔도 됩니다 😄 
너무 잘하고 있습니다... 든든하다👍

choiyoubin commented:
Compositional Layout을 두번째 쓰는거라 익숙하지 않았는데, switch 문으로 뺴고 보니 깔끔하게 구현할 수 있었던 것 같습니다.
추후 Cell 작업시 더 신중하게 작업해보겠습니당!
